### PR TITLE
Add MDS047 rule for detecting ambiguous emphasis sequences

### PR DESCRIPTION
**Pull request by jeduden**
## Summary

Implements MDS047, a new linting rule that detects emphasis sequences whose meaning is ambiguous or difficult to predict at a glance. The rule identifies three problematic patterns in Markdown source: long delimiter runs, backslash-escaped delimiters adjacent to runs, and repeated same-delimiter sequences separated by non-whitespace.

## Key Changes

- **New rule implementation** (`internal/rules/ambiguousemphasis/rule.go`):
  - Detects three patterns: delimiter runs exceeding a configurable maximum length, escaped delimiters glued to unescaped runs (`*\*`, `_\_`), and three or more same-delimiter runs on a line separated only by non-whitespace
  - Properly handles code span and code block exclusion to avoid false positives
  - Implements deduplication per `(char, length)` to collapse symmetric opener/closer pairs into single diagnostics
  - Configurable via three settings: `max-run` (int), `forbid-escaped-in-run` (bool), `forbid-adjacent-same-delim` (bool)
  - Disabled by default with no-op configuration

- **Comprehensive test coverage** (`internal/rules/ambiguousemphasis/rule_test.go`):
  - Tests for each detector pattern
  - Verification that code spans and code blocks suppress diagnostics
  - Edge case handling (escaped backslashes, whitespace separation, deduplication)
  - Settings validation and type checking

- **Documentation** (`internal/rules/MDS047-ambiguous-emphasis/README.md`):
  - Detailed explanation of the three problematic patterns
  - Configuration guide with all three settings
  - Examples of bad patterns and good alternatives
  - Edge case documentation

- **Test fixtures** (bad/ and good/ directories):
  - Fixture files for long runs, escaped delimiters, adjacent same-delimiters, and the "Peter Piper" example from the original rant
  - Good examples showing code span and fenced block suppression

- **Integration**:
  - Registered rule in `cmd/mdsmith/main.go` and test files
  - Updated plan 111 status to complete
  - Added to rules index

## Notable Implementation Details

- **Source-level scanning**: Operates on raw bytes rather than AST to catch patterns before CommonMark resolution
- **Code span masking**: Replaces code span content with spaces to exclude from delimiter accounting while preserving column positions
- **Escape handling**: Properly distinguishes between escaped delimiters (which don't form runs) and unescaped ones
- **Whitespace-aware gap detection**: Treats gaps containing spaces as clean separations since CommonMark's flanking rules resolve them unambiguously
- **Deduplication strategy**: Emits one diagnostic per unique `(char, length)` per line, so `***x***` (two three-star runs) produces a single report

https://claude.ai/code/session_01FMiFqJgrmiitAod8Y7cejG

### CODE REVIEW

**Comment by jeduden** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/jeduden/mdsmith/pull/195#pullrequestreview-4190924925)

File: internal/rules/ambiguousemphasis/rule.go [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+					Message:  "adjacent same-delimiter emphasis is ambiguous",
+				})
+				break
+			}
+			if emitted[k] {
+				break
+			}
+		}
+		_ = found2
+	}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/ambiguousemphasis/rule.go]: Fixed in d49a8cb — removed `found2` and the `_ = found2` placeholder.

File: internal/rules/ambiguousemphasis/rule.go [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+	for i := 0; i < len(f.Source); i++ {
+		if f.Source[i] == '\n' {
+			count++
+			if count == lineNum {
+				return i + 1
+			}
+		}
+	}
+	return len(f.Source)
+}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/ambiguousemphasis/rule.go]: Fixed in d49a8cb — replaced the per-line `lineStartOffset` scan with a single `computeLineStarts` precompute that `maskCodeSpans` indexes into.

File: internal/rules/ambiguousemphasis/rule.go [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+					RuleName: r.Name(),
+					Severity: lint.Warning,
+					Message:  "adjacent same-delimiter emphasis is ambiguous",
+				})
+				break
+			}
+			if emitted[k] {
+				break
+			}
+		}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/ambiguousemphasis/rule.go]: Fixed in dd3443e/08ae719 — `adjacentSameDelimDiags` is now the suggested single-pass O(n) loop using an `adjacentRunState` map keyed by (char, length).

File: go.mod [hunk 4]
```diff
@@ -15,6 +15,7 @@ require (
 	github.com/neurosnap/sentences v1.1.2
 	github.com/spf13/pflag v1.0.10
 	github.com/stretchr/testify v1.11.1
+	github.com/tetratelabs/wazero v1.11.0
 	github.com/yuin/goldmark v1.8.2
 	go.abhg.dev/goldmark/frontmatter v0.3.0
```
[the remark this answers is not in the document]
Comment (jeduden) [reply go.mod:20]: Fixed in d5cd133 — reverted wazero to indirect; only build-tagged spike code references it.

File: internal/rules/ambiguousemphasis/rule.go [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+)
+
+func init() {
+	rule.Register(&Rule{})
+}
+
+// Rule flags ambiguous emphasis sequences. Each detector is gated by
+// its own setting and contributes at most one diagnostic per
+// (char, length) shape per line, so symmetric openers and closers
+// collapse into a single report.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/ambiguousemphasis/rule.go]: Fixed in d5cd133 — reworded the Rule type comment to describe the actual per-detector behavior: long-run and adjacent-same-delim dedupe per (char, length); escaped-in-run emits one diagnostic per matching escape.

File: internal/rules/ambiguousemphasis/rule.go [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+	}
+	return true
+}
+
+// codeSpanRange records the inclusive byte range of one CodeSpan's
+// delimited content within f.Source.
+type codeSpanRange struct {
+	start int // absolute byte offset in source, inclusive
+	end   int // absolute byte offset in source, exclusive
+}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/ambiguousemphasis/rule.go]: Fixed in d5cd133 — codeSpanRange doc now reads "half-open [start, end) byte range", matching the field comments.

File: internal/rules/ambiguousemphasis/rule.go [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+	var diags []lint.Diagnostic
+	for _, e := range escapes {
+		matched := false
+		for _, run := range runs {
+			if run.char == e.char && run.end == e.pos {
+				matched = true
+				break
+			}
+		}
+		if !matched {
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/ambiguousemphasis/rule.go]: Fixed in d5cd133 — escapedInRunDiags now indexes runs by (char, end) for O(1) lookups per escape, making the rule linear-time on lines with many delimiters.

File: internal/rules/MDS047-ambiguous-emphasis/good/in-code-span.md [hunk 8]
```diff
@@ -0,0 +1,10 @@
+---
+settings:
+  max-run: 2
+  forbid-escaped-in-run: true
+  forbid-adjacent-same-delim: true
+---
+# In code span
+
+The string `*****\*a*` inside an inline code span must not flag, and
+neither must `__a__b__` nor `***Peter* Piper**` when wrapped this way.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS047-ambiguous-emphasis/good/in-code-span.md:10]: Looks like a false positive — line 10 is 70 characters (`awk 'NR==10 {print length}'` confirms), well under the MDS001 default of 80. The integration test `TestRuleFixtures/MDS047/good/in-code-span` runs through `checkAllRules` and passes locally. Leaving the line as is.

File: internal/rules/ambiguousemphasis/rule.go [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+func gapNonEmptyAllNonWhitespace(b []byte) bool {
+	if len(b) == 0 {
+		return false
+	}
+	for _, c := range b {
+		if c == ' ' || c == '\t' {
+			return false
+		}
+	}
+	return true
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/ambiguousemphasis/rule.go:311]: Fixed in b8573a0 — clarified the docstring to state "ASCII space or tab" explicitly and explain why this covers the relevant CommonMark whitespace cases (gaps live within a single split line, so newlines never appear).

**Comment by jeduden** [comment]
🟢 **Merge Queue** — picked up

This PR is in the queue and will be batched with other `queue`-labelled PRs.

**Next:** No action needed — you'll get another comment when CI starts on the batch. [View merge queue run](https://github.com/jeduden/mdsmith/actions/runs/25248960437).

**Comment by jeduden** [comment]
⏳ **Merge Queue** — requeued

The merge queue hit a transient error while processing this PR:

> batch creation failed: merging PR #195: git merge failed (exit 2): error: Your local changes to the following files would be overwritten by merge: PLAN.md internal/rules/index.md Please commit your changes or stash them be…

[View merge queue run](https://github.com/jeduden/mdsmith/actions/runs/25248960437).

**Next:** No action needed — the queue will retry automatically on the next run.

**Comment by jeduden** [comment]
⏳ **Merge Queue** — requeued

The merge queue hit a transient error while processing this PR:

> merging PR #206: git merge failed (exit 2): error: Your local changes to the following files would be overwritten by merge: PLAN.md Please commit your changes or stash them before you merge. Aborting…

[View merge queue run](https://github.com/jeduden/mdsmith/actions/runs/25249013389).

**Next:** No action needed — the queue will retry automatically on the next run.

**Comment by jeduden** [comment]
🟢 **Merge Queue** — picked up

This PR is in the queue and will be batched with other `queue`-labelled PRs.

**Next:** No action needed — you'll get another comment when CI starts on the batch. [View merge queue run](https://github.com/jeduden/mdsmith/actions/runs/25249184997).

**Comment by jeduden** [comment]
🔵 **Merge Queue** — CI running

Merged into batch branch [`merge-queue/batch-195-1777715156`](https://github.com/jeduden/mdsmith/tree/merge-queue/batch-195-1777715156). [View CI run](https://github.com/jeduden/mdsmith/actions/runs/25249190198).

**Next:** No action needed — you'll be notified when CI completes.

**Comment by jeduden** [comment]
❌ **Merge Queue** — CI failed

The [batch CI run](https://github.com/jeduden/mdsmith/actions/runs/25249190198) failed with this PR in it.

**Next:** Fix the failure, push updates, then re-add the `queue` label to retry.

**Comment by jeduden** [comment]
🟢 **Merge Queue** — picked up

This PR is in the queue and will be batched with other `queue`-labelled PRs.

**Next:** No action needed — you'll get another comment when CI starts on the batch. [View merge queue run](https://github.com/jeduden/mdsmith/actions/runs/25249594854).

**Comment by jeduden** [comment]
🔵 **Merge Queue** — CI running

Merged into batch branch [`merge-queue/batch-195-1777716638`](https://github.com/jeduden/mdsmith/tree/merge-queue/batch-195-1777716638). [View CI run](https://github.com/jeduden/mdsmith/actions/runs/25249598595).

**Next:** No action needed — you'll be notified when CI completes.

**Comment by jeduden** [comment]
🟢 **Merge Queue** — picked up

This PR is in the queue and will be batched with other `queue`-labelled PRs.

**Next:** No action needed — you'll get another comment when CI starts on the batch. [View merge queue run](https://github.com/jeduden/mdsmith/actions/runs/25249609764).

**Comment by jeduden** [comment]
🔵 **Merge Queue** — CI running

Merged into batch branch [`merge-queue/batch-195-1777716692`](https://github.com/jeduden/mdsmith/tree/merge-queue/batch-195-1777716692). [View CI run](https://github.com/jeduden/mdsmith/actions/runs/25249613569).

**Next:** No action needed — you'll be notified when CI completes.

**Comment by jeduden** [comment]
🔍 **Merge Queue** — bisecting

A larger batch failed CI. Bisection is isolating the culprit: this run tests up to **2 of 4** candidate PRs on [`merge-queue/batch-bisect-195-1777748428`](https://github.com/jeduden/mdsmith/tree/merge-queue/batch-bisect-195-1777748428). [View current bisect CI run](https://github.com/jeduden/mdsmith/actions/runs/25259441999).

**Next:** No action needed — you'll be notified when the culprit is isolated or this PR merges.

**Comment by jeduden** [comment]
✅ **Merge Queue** — merged

This PR landed on `main` via commit [`43dfa16`](https://github.com/jeduden/mdsmith/commit/43dfa16566f25f1d318d2c9b505876946e970afe). [CI run that validated the merge](https://github.com/jeduden/mdsmith/actions/runs/25259441999).

**Next:** Done — nothing more to do here.
